lib/spack/spack/spec.py

```diff
@@ -4475,6 +4477,43 @@ def cformat(self, *args, **kwargs):
         kwargs.setdefault("color", None)
         return self.format(*args, **kwargs)
 
+    def format_path(self, format_string: str):
+        """Given a `format_string` that is intended as a path, generate a string
+        like from `Spec.format`, but eliminate extra path separators introduced by
+        formatting of Spec properties.
+
+        Path separators explicitly added to the string are preserved, so for example
+        "{name}/{version}" would generate a directory based on the Spec's name, and
+        a subdirectory based on its version; this function guarantees though that
+        the resulting string would only have two directories (i.e. that if under
+        normal circumstances that `str(Spec.version)` would contain a path
+        separator, it would not in this case).
+
+        Note that POSIX or Windows-style paths can be supplied on either POSIX or
+        Windows, and both "/" and "\" are treated as path separators (as a side
+        effect, "invalid" paths which mix these will be converted to OS-appropriate
+        paths with consistent separators).
+        """
+        format_component_with_sep = r"\{[^}]*[/\\][^}]*}"
+        if re.search(format_component_with_sep, format_string):
+            raise SpecFormatPathError(
+                f"Invalid path format string: cannot contain {{/...}}\n\t{format_string}"
+            )
+
+        root = ""
+        extraction_tuple = _extract_root(format_string)
+        if extraction_tuple:
+            root, format_string = extraction_tuple
+
+        # If we want to think of a string like "a/b/c" as a path (with 3 subdirs)
+        # on Windows, we cannot use pathlib (since "/" is not a path separator
```

Comment:
I don't get this comment

```
In [1]: pathlib.PureWindowsPath("abc/def").parts
Out[1]: ('abc', 'def')
```

Reply:
Thanks, this was wrong. Although:

```
>>> y = pathlib.PurePosixPath(r"a\b\c")
>>> y.parts
('a\\b\\c',)
```

and furthermore I don't want to use pathlib for reasons discussed in https://github.com/spack/spack/pull/39398#discussion_r1301244744 (and I updated the comment accordingly after correcting it).

Reply:
I don't think I have to explain to you that backslash is indeed not a path separator on posix :confused:

Reply:
You are correct that you don't have to explain that to me: until we forbid `\` in our config files etc. (a proposal elsewhere which I like, but has not taken effect yet), I think we should handle `\` sensibly on Linux.

Overall, constructing a `PurePath` has a system-dependent behavior which is not obvious (this concern is in addition to how `pathlib`/`os.path` handles invalid path characters).